sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -4588,6 +4588,45 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
       sql("SELECT /*+ hash(t2) */ * FROM t1 join t2 on c1 = c2")
     }
   }
+
+  test("SPARK-41538: Metadata column should be appended at the end of project") {
+    val tableName = "table_1"
+    val viewName = "view_1"
+    withTable(tableName) {
+      withView(viewName) {
+        sql(s"CREATE TABLE $tableName (a ARRAY<STRING>, s STRUCT<id: STRING>) USING parquet")
+        val id = "id1"
+        sql(s"INSERT INTO $tableName values(ARRAY('a'), named_struct('id', '$id'))")
+        sql(
+          s"""
+             |CREATE VIEW $viewName (id)
+             |AS WITH source AS (
+             |    SELECT * FROM $tableName
+             |),
+             |renamed AS (
+             |    SELECT s.id FROM source
+             |)
+             |SELECT id FROM renamed
+             |""".stripMargin)
+        val query =
+          s"""
+             |with foo AS (
+             |  SELECT '$id' as id
+             |),
+             |bar AS (
+             |  SELECT '$id' as id
+             |)
+             |SELECT
+             |  1
+             |FROM foo
+             |FULL OUTER JOIN bar USING(id)
```

Comment:
The full outer join will trigger the rule `AddMetadataColumns` to add meta columns over the whole plan.
Before rule `AddMetadataColumns`:
```
!Filter isnotnull(id#223)
+- Project [coalesce(id#227, id#225) AS id#228]
   +- Join FullOuter, (id#227 = id#225)
      :- Project [coalesce(id#223, id#224) AS id#227]
      :  +- Join FullOuter, (id#223 = id#224)
      :     :- SubqueryAlias foo
      :     :  +- CTERelationRef 2, true, [id#223]
      :     +- SubqueryAlias bar
      :        +- CTERelationRef 3, true, [id#224]
      +- SubqueryAlias spark_catalog.default.view_1
         +- View (`spark_catalog`.`default`.`view_1`, [id#225])
            +- Project [cast(id#226 as string) AS id#225]
               +- WithCTE
                  :- CTERelationDef 4, false
                  :  +- SubqueryAlias source
                  :     +- Project [a#218, s#219]
                  :        +- SubqueryAlias spark_catalog.default.table_1
                  :           +- Relation spark_catalog.default.table_1[a#218,s#219] parquet
                  :- CTERelationDef 5, false
                  :  +- SubqueryAlias renamed
                  :     +- Project [s#219.id AS id#226]
                  :        +- SubqueryAlias source
                  :           +- CTERelationRef 4, true, [a#218, s#219]
                  +- Project [id#226]
                     +- SubqueryAlias renamed
                        +- CTERelationRef 5, true, [id#226]

```
After the rule:
```
Filter isnotnull(id#223)
+- Project [id#227, id#225, id#223, id#224, coalesce(id#227, id#225) AS id#228]
   +- Join FullOuter, (id#227 = id#225)
      :- Project [id#223, id#224, coalesce(id#223, id#224) AS id#227]
      :  +- Join FullOuter, (id#223 = id#224)
      :     :- SubqueryAlias foo
      :     :  +- CTERelationRef 2, true, [id#223]
      :     +- SubqueryAlias bar
      :        +- CTERelationRef 3, true, [id#224]
      +- SubqueryAlias spark_catalog.default.view_1
         +- View (`spark_catalog`.`default`.`view_1`, [id#225])
            +- Project [cast(id#226 as string) AS id#225]
               +- WithCTE
                  :- CTERelationDef 4, false
                  :  +- SubqueryAlias source
                  :     +- Project [_metadata#221, a#218, s#219]
                  :        +- SubqueryAlias spark_catalog.default.table_1
                  :           +- Relation spark_catalog.default.table_1[a#218,s#219,_metadata#221] parquet
                  :- CTERelationDef 5, false
                  :  +- SubqueryAlias renamed
                  :     +- Project [s#219.id AS id#226]
                  :        +- SubqueryAlias source
                  :           +- CTERelationRef 4, true, [a#218, s#219]
                  +- Project [id#226]
                     +- SubqueryAlias renamed
                        +- CTERelationRef 5, true, [id#226]
```